File: packages/typescript-sdk/src/xaa.ts
/**
 * Cross-App Access (XAA) client helpers — TypeScript port of the Go SDK's
 * xaa.go. Implements the two legs of the IETF
 * draft-ietf-oauth-identity-assertion-authz-grant flow:
 *
 *   1. requestIdJag()       — Token Exchange at the IdP (RFC 8693).
 *   2. exchangeForAccessToken() — jwt-bearer grant at the Resource AS (RFC 7523).
 *   3. crossAppAccess()     — convenience wrapper that chains both.
 *
 * The functions are intentionally thin — no caching, no automatic refresh.
 * Callers manage token lifecycle themselves. The point is to make the wire
 * format hard to get wrong, not to hide it.
 */

// Standard grant + token type URNs.
export const GRANT_TYPE_TOKEN_EXCHANGE =
  'urn:ietf:params:oauth:grant-type:token-exchange';
export const GRANT_TYPE_JWT_BEARER =
  'urn:ietf:params:oauth:grant-type:jwt-bearer';
export const TOKEN_TYPE_ID_JAG =
  'urn:ietf:params:oauth:token-type:id-jag';
export const TOKEN_TYPE_ID_TOKEN =
  'urn:ietf:params:oauth:token-type:id_token';
export const TOKEN_TYPE_REFRESH_TOKEN =
  'urn:ietf:params:oauth:token-type:refresh_token';

/** Input to {@link requestIdJag}. Mirrors the Token Exchange form fields. */
export interface RequestIdJagInput {
  /**
   * IdP AS's `/idjag/token` endpoint. For AuthSec today:
   * `https://<authsec>/authsec/oauth/v2/idjag/token`.
   */
  idpTokenEndpoint: string;

  /** Requesting client's credentials (xaa_client_apps row for AuthSec). */
  clientId: string;
  clientSecret: string;

  /** The user's id_token (or refresh token) — their identity assertion. */
  subjectToken: string;

  /** Defaults to TOKEN_TYPE_ID_TOKEN if omitted. */
  subjectTokenType?: string;

  /**
   * Resource AS issuer URI the ID-JAG will be redeemed at. Required.
   * For AuthSec where IdP and Resource AS share a host, this is the AuthSec
   * issuer (e.g. `https://prod.api.authsec.ai`).
   */
  audience: string;

  /**
   * RFC 8707 target resource URI — the specific MCP the access token will be
   * valid for. Required by AuthSec because XAA policy keys on
   * resource_server_id.
   */
  resource: string;

  /** Desired scope set. Empty = whatever the IdP's policy allows. */
  scopes?: string[];

  /**
   * Optional fetch implementation override. Defaults to globalThis.fetch.
   * Useful for injecting timeouts / retries / Node 18- shims.
   */
  fetch?: typeof fetch;
}

/** IdP response per RFC 8693 §2.2.1. */
export interface IdJagResponse {
  access_token: string;
  issued_token_type: string;
  /** "N_A" for ID-JAGs per the spec — they're not bearer access tokens. */
  token_type: string;
  expires_in: number;
  scope?: string;
}

/**
 * Typed OAuth error per RFC 6749 §5.2. Use `instanceof OAuthError` (or
 * `err.code === 'invalid_grant'` etc.) to branch on the error code.
 */
export class OAuthError extends Error {
  readonly httpStatus: number;
  readonly code: string;
  readonly description?: string;
  readonly uri?: string;

  constructor(httpStatus: number, body: unknown) {
    const obj = (body && typeof body === 'object' ? body : {}) as Record<string, unknown>;
    const code = typeof obj.error === 'string' ? obj.error : `http_${httpStatus}`;
    const description = typeof obj.error_description === 'string' ? obj.error_description : undefined;
    super(description ? `oauth ${code}: ${description}` : `oauth ${code}`);
    this.name = 'OAuthError';
    this.httpStatus = httpStatus;
    this.code = code;
    this.description = description;
    this.uri = typeof obj.error_uri === 'string' ? obj.error_uri : undefined;
  }
}

/**
 * Exchange a user's identity assertion at the IdP for an ID-JAG.
 *
 * @example
 * ```ts
 * const idjag = await requestIdJag({
 *   idpTokenEndpoint: 'https://prod.api.authsec.ai/authsec/oauth/v2/idjag/token',
 *   clientId: 'my-agent',
 *   clientSecret: process.env.AGENT_SECRET!,
 *   subjectToken: userIdToken,
 *   audience: 'https://prod.api.authsec.ai',
 *   resource: 'https://my-mcp.example.com/mcp',
 *   scopes: ['orders:read'],
 * });
 * ```
 */
export async function requestIdJag(
  input: RequestIdJagInput,
): Promise<IdJagResponse> {
  validateRequestIdJagInput(input);
  const body = new URLSearchParams();
  body.set('grant_type', GRANT_TYPE_TOKEN_EXCHANGE);
  body.set('requested_token_type', TOKEN_TYPE_ID_JAG);
  body.set('subject_token', input.subjectToken);
  body.set('subject_token_type', input.subjectTokenType ?? TOKEN_TYPE_ID_TOKEN);
  body.set('audience', input.audience);
  body.set('resource', input.resource);
  if (input.scopes && input.scopes.length > 0) {
    body.set('scope', input.scopes.join(' '));
  }

  const f = input.fetch ?? fetch;
  const resp = await f(input.idpTokenEndpoint, {
    method: 'POST',
    body,
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
      'Accept': 'application/json',
      'Authorization':
        'Basic ' + btoa(`${input.clientId}:${input.clientSecret}`),
    },
  });
  const text = await resp.text();
  if (!resp.ok) {
    throw new OAuthError(resp.status, safeJSON(text));
  }
  const parsed = safeJSON(text) as IdJagResponse;
  if (!parsed?.access_token) {
    throw new Error('idjag exchange: empty access_token in response');
  }
  return parsed;
}

/** Input to {@link exchangeForAccessToken}. */
export interface ExchangeForAccessTokenInput {
  /**
   * Resource AS's `/token` endpoint. For AuthSec the same host as the IdP:
   * `https://<authsec>/authsec/oauth/v2/token`.
   */
  resourceAsTokenEndpoint: string;

  /** The ID-JAG returned from {@link requestIdJag}. */
  idJag: string;

  /** Optional narrowing of the ID-JAG's scope set. */
  scopes?: string[];

  /**
   * Optional client auth at the Resource AS. The ID-JAG already carries a
   * client_id claim, so most Resource AS deployments don't require this.
   */
  clientId?: string;
  clientSecret?: string;

  fetch?: typeof fetch;
}

/** Standard OAuth 2.0 access token response (RFC 6749 §5.1). */
export interface AccessTokenResponse {
  access_token: string;
  token_type: string;
  expires_in: number;
  scope?: string;
  refresh_token?: string;
}

/** Redeem an ID-JAG at the Resource AS for a usable access token. */
export async function exchangeForAccessToken(
  input: ExchangeForAccessTokenInput,
): Promise<AccessTokenResponse> {
  if (!input.resourceAsTokenEndpoint?.trim()) {
    throw new Error('resourceAsTokenEndpoint required');
  }
  if (!input.idJag?.trim()) {
    throw new Error('idJag required');
  }

  const body = new URLSearchParams();
  body.set('grant_type', GRANT_TYPE_JWT_BEARER);
  body.set('assertion', input.idJag);
  if (input.scopes && input.scopes.length > 0) {
    body.set('scope', input.scopes.join(' '));
  }

  const headers: Record<string, string> = {
    'Content-Type': 'application/x-www-form-urlencoded',
    'Accept': 'application/json',
  };
  if (input.clientId) {
    headers['Authorization'] =
      'Basic ' + btoa(`${input.clientId}:${input.clientSecret ?? ''}`);
  }

  const f = input.fetch ?? fetch;
  const resp = await f(input.resourceAsTokenEndpoint, {
    method: 'POST',
    body,
    headers,
  });
  const text = await resp.text();
  if (!resp.ok) {
    throw new OAuthError(resp.status, safeJSON(text));
  }
  const parsed = safeJSON(text) as AccessTokenResponse;
  if (!parsed?.access_token) {
    throw new Error('jwt-bearer exchange: empty access_token in response');
  }
  return parsed;
}

/** Input to {@link crossAppAccess} — both endpoints + everything in one call. */
export interface CrossAppAccessInput {
  idpTokenEndpoint: string;
  resourceAsTokenEndpoint: string;
  clientId: string;
  clientSecret: string;
  subjectToken: string;
  subjectTokenType?: string;
  audience: string;
  resource: string;
  scopes?: string[];
  fetch?: typeof fetch;
}

/**
 * Run the full two-leg XAA flow: IdP → ID-JAG → Resource AS → access token.
 * Use the granular {@link requestIdJag} / {@link exchangeForAccessToken} when
 * you want to inspect or cache the ID-JAG between calls.
 */
export async function crossAppAccess(
  input: CrossAppAccessInput,
): Promise<AccessTokenResponse> {
  const idjag = await requestIdJag({
    idpTokenEndpoint: input.idpTokenEndpoint,
    clientId: input.clientId,
    clientSecret: input.clientSecret,
    subjectToken: input.subjectToken,
    subjectTokenType: input.subjectTokenType,
    audience: input.audience,
    resource: input.resource,
    scopes: input.scopes,
    fetch: input.fetch,
  });
  return exchangeForAccessToken({
    resourceAsTokenEndpoint: input.resourceAsTokenEndpoint,
    idJag: idjag.access_token,
    scopes: input.scopes,
    fetch: input.fetch,
  });
}

// ── helpers ─────────────────────────────────────────────────────────────────

function validateRequestIdJagInput(in_: RequestIdJagInput): void {
  if (!in_.idpTokenEndpoint?.trim()) throw new Error('idpTokenEndpoint required');
  if (!in_.clientId?.trim()) throw new Error('clientId required');
  if (!in_.subjectToken?.trim()) throw new Error('subjectToken required');
  if (!in_.audience?.trim()) throw new Error('audience required');
  if (!in_.resource?.trim()) throw new Error('resource required');
}

function safeJSON(text: string): unknown {
  if (!text) return null;
  try {
    return JSON.parse(text);
  } catch {
    return null;
  }
}

File: packages/go-sdk/xaa.go
package authsec

// Cross-App Access (XAA) client helpers.
//
// These wrap the IETF draft-ietf-oauth-identity-assertion-authz-grant flow so
// an agent or chained MCP server can:
//
//	1. Exchange a user's id_token at the IdP for an ID-JAG via
//	   POST /authsec/oauth/v2/idjag/token (Token Exchange, RFC 8693).
//	2. Redeem the ID-JAG at the Resource AS for an access token via
//	   POST /authsec/oauth/v2/token (jwt-bearer grant, RFC 7523).
//
// The two functions in this file are thin HTTP wrappers — no hidden state,
// no caching. Callers manage token reuse + refresh themselves. The point is
// to make the spec wire-format hard to get wrong, not to hide it.

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// Grant + token type constants (IETF draft-ietf-oauth-identity-assertion-authz-grant + RFC 8693).
const (
	GrantTypeTokenExchange = "urn:ietf:params:oauth:grant-type:token-exchange"
	GrantTypeJWTBearer     = "urn:ietf:params:oauth:grant-type:jwt-bearer"
	TokenTypeIDJAG         = "urn:ietf:params:oauth:token-type:id-jag"
	TokenTypeIDToken       = "urn:ietf:params:oauth:token-type:id_token"
	TokenTypeRefreshToken  = "urn:ietf:params:oauth:token-type:refresh_token"
)

// RequestIDJAGInput is the input for RequestIDJAG. Mirrors the Token Exchange
// form parameters one-to-one so callers reading the IETF draft see the same
// names.
type RequestIDJAGInput struct {
	// IdPTokenEndpoint is the IdP AS's /idjag/token URL. For AuthSec
	// today: https://<authsec>/authsec/oauth/v2/idjag/token.
	IdPTokenEndpoint string

	// ClientID + ClientSecret authenticate the requesting agent at the
	// IdP. Maps to the xaa_client_apps row when AuthSec is the IdP.
	ClientID     string
	ClientSecret string

	// SubjectToken is the user's identity assertion — usually the
	// id_token they got from their original login at the IdP. Refresh
	// tokens are also accepted by the spec but most IdPs prefer id_token
	// here.
	SubjectToken string

	// SubjectTokenType. Defaults to TokenTypeIDToken when empty.
	SubjectTokenType string

	// Audience is the Resource AS issuer the ID-JAG will be redeemed at.
	// REQUIRED. For internal AuthSec deployments this is the AuthSec
	// issuer URL itself (e.g. https://prod.api.authsec.ai) since AuthSec
	// is also the Resource AS.
	Audience string

	// Resource is the RFC 8707 target resource URI — the specific MCP
	// the access token will be valid for. REQUIRED by AuthSec because
	// XAA policy is keyed on resource_server_id.
	Resource string

	// Scopes is the desired scope set. Empty = "give me whatever the
	// IdP policy permits for this (client, resource)".
	Scopes []string

	// HTTPClient lets callers inject timeouts, proxies, retries. nil =
	// a default client with a 10s timeout.
	HTTPClient *http.Client
}

// IDJAGResponse is the IdP's RFC 8693 §2.2.1 reply.
type IDJAGResponse struct {
	AccessToken     string `json:"access_token"`
	IssuedTokenType string `json:"issued_token_type"`
	TokenType       string `json:"token_type"` // "N_A" per spec for ID-JAG
	ExpiresIn       int    `json:"expires_in"`
	Scope           string `json:"scope,omitempty"`
}

// OAuthError is the RFC 6749 error response we surface to callers as a typed
// error. errors.As lets callers branch on the error code without parsing
// strings.
type OAuthError struct {
	HTTPStatus       int    `json:"-"`
	Code             string `json:"error"`
	Description      string `json:"error_description,omitempty"`
	URI              string `json:"error_uri,omitempty"`
}

func (e *OAuthError) Error() string {
	if e.Description != "" {
		return fmt.Sprintf("oauth %s: %s", e.Code, e.Description)
	}
	return "oauth " + e.Code
}

// RequestIDJAG performs the Token Exchange at the IdP and returns the ID-JAG.
//
// Typical use:
//
//	idjag, err := authsec.RequestIDJAG(ctx, authsec.RequestIDJAGInput{
//	    IdPTokenEndpoint: "https://prod.api.authsec.ai/authsec/oauth/v2/idjag/token",
//	    ClientID:         "my-agent",
//	    ClientSecret:     os.Getenv("AGENT_SECRET"),
//	    SubjectToken:     userIDToken,
//	    Audience:         "https://prod.api.authsec.ai",
//	    Resource:         "https://my-mcp.example.com/mcp",
//	    Scopes:           []string{"orders:read"},
//	})
func RequestIDJAG(ctx context.Context, in RequestIDJAGInput) (*IDJAGResponse, error) {
	if err := in.validate(); err != nil {
		return nil, err
	}
	form := url.Values{}
	form.Set("grant_type", GrantTypeTokenExchange)
	form.Set("requested_token_type", TokenTypeIDJAG)
	form.Set("subject_token", in.SubjectToken)
	subjType := in.SubjectTokenType
	if subjType == "" {
		subjType = TokenTypeIDToken
	}
	form.Set("subject_token_type", subjType)
	form.Set("audience", in.Audience)
	form.Set("resource", in.Resource)
	if len(in.Scopes) > 0 {
		form.Set("scope", strings.Join(in.Scopes, " "))
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, in.IdPTokenEndpoint, strings.NewReader(form.Encode()))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Accept", "application/json")
	req.SetBasicAuth(in.ClientID, in.ClientSecret)

	resp, err := defaultClient(in.HTTPClient).Do(req)
	if err != nil {
		return nil, fmt.Errorf("idjag exchange: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	if err != nil {
		return nil, fmt.Errorf("idjag exchange: read body: %w", err)
	}
	if resp.StatusCode != http.StatusOK {
		return nil, parseOAuthError(resp.StatusCode, body)
	}
	var out IDJAGResponse
	if err := json.Unmarshal(body, &out); err != nil {
		return nil, fmt.Errorf("idjag exchange: decode response: %w", err)
	}
	if out.AccessToken == "" {
		return nil, errors.New("idjag exchange: empty access_token in response")
	}
	return &out, nil
}

func (in RequestIDJAGInput) validate() error {
	switch {
	case strings.TrimSpace(in.IdPTokenEndpoint) == "":
		return errors.New("IdPTokenEndpoint required")
	case strings.TrimSpace(in.ClientID) == "":
		return errors.New("ClientID required")
	case strings.TrimSpace(in.SubjectToken) == "":
		return errors.New("SubjectToken required")
	case strings.TrimSpace(in.Audience) == "":
		return errors.New("Audience required")
	case strings.TrimSpace(in.Resource) == "":
		return errors.New("Resource required")
	}
	return nil
}

// ExchangeForAccessTokenInput drives the second leg — redeeming the ID-JAG at
// the Resource AS for a usable access token.
type ExchangeForAccessTokenInput struct {
	// ResourceASTokenEndpoint is the Resource AS's /token endpoint. For
	// AuthSec this is the same host as the IdP because AuthSec serves
	// both roles in v1: https://<authsec>/authsec/oauth/v2/token.
	ResourceASTokenEndpoint string

	// IDJAG is the assertion received from RequestIDJAG().
	IDJAG string

	// Scopes optionally narrows the access token's scope to a subset of
	// what the ID-JAG carries. Empty = take everything the IDJAG offers.
	Scopes []string

	// ClientID + ClientSecret are OPTIONAL at the Resource AS — the ID-JAG
	// already carries a client_id claim. Supplied here only if your
	// Resource AS requires client authentication on the JWT-bearer grant.
	ClientID     string
	ClientSecret string

	HTTPClient *http.Client
}

// AccessTokenResponse is the standard OAuth token response (RFC 6749 §5.1).
type AccessTokenResponse struct {
	AccessToken  string `json:"access_token"`
	TokenType    string `json:"token_type"`
	ExpiresIn    int    `json:"expires_in"`
	Scope        string `json:"scope,omitempty"`
	RefreshToken string `json:"refresh_token,omitempty"`
}

// ExchangeForAccessToken redeems an ID-JAG for an access token at the Resource
// AS via the RFC 7523 jwt-bearer grant.
func ExchangeForAccessToken(ctx context.Context, in ExchangeForAccessTokenInput) (*AccessTokenResponse, error) {
	switch {
	case strings.TrimSpace(in.ResourceASTokenEndpoint) == "":
		return nil, errors.New("ResourceASTokenEndpoint required")
	case strings.TrimSpace(in.IDJAG) == "":
		return nil, errors.New("IDJAG required")
	}

	form := url.Values{}
	form.Set("grant_type", GrantTypeJWTBearer)
	form.Set("assertion", in.IDJAG)
	if len(in.Scopes) > 0 {
		form.Set("scope", strings.Join(in.Scopes, " "))
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, in.ResourceASTokenEndpoint, strings.NewReader(form.Encode()))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Accept", "application/json")
	if in.ClientID != "" {
		req.SetBasicAuth(in.ClientID, in.ClientSecret)
	}

	resp, err := defaultClient(in.HTTPClient).Do(req)
	if err != nil {
		return nil, fmt.Errorf("jwt-bearer exchange: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	if err != nil {
		return nil, fmt.Errorf("jwt-bearer exchange: read body: %w", err)
	}
	if resp.StatusCode != http.StatusOK {
		return nil, parseOAuthError(resp.StatusCode, body)
	}
	var out AccessTokenResponse
	if err := json.Unmarshal(body, &out); err != nil {
		return nil, fmt.Errorf("jwt-bearer exchange: decode response: %w", err)
	}
	if out.AccessToken == "" {
		return nil, errors.New("jwt-bearer exchange: empty access_token in response")
	}
	return &out, nil
}

// CrossAppAccess is the convenience wrapper most agents will actually use —
// chains the two steps so the caller doesn't have to wire them. Use the
// granular RequestIDJAG / ExchangeForAccessToken when you need to inspect or
// cache the ID-JAG between calls.
type CrossAppAccessInput struct {
	IdPTokenEndpoint        string
	ResourceASTokenEndpoint string
	ClientID                string
	ClientSecret            string
	SubjectToken            string
	SubjectTokenType        string
	Audience                string
	Resource                string
	Scopes                  []string
	HTTPClient              *http.Client
}

// CrossAppAccess runs the full two-leg flow and returns the final access
// token. Errors propagate from whichever leg fails.
func CrossAppAccess(ctx context.Context, in CrossAppAccessInput) (*AccessTokenResponse, error) {
	idjag, err := RequestIDJAG(ctx, RequestIDJAGInput{
		IdPTokenEndpoint: in.IdPTokenEndpoint,
		ClientID:         in.ClientID,
		ClientSecret:     in.ClientSecret,
		SubjectToken:     in.SubjectToken,
		SubjectTokenType: in.SubjectTokenType,
		Audience:         in.Audience,
		Resource:         in.Resource,
		Scopes:           in.Scopes,
		HTTPClient:       in.HTTPClient,
	})
	if err != nil {
		return nil, err
	}
	return ExchangeForAccessToken(ctx, ExchangeForAccessTokenInput{
		ResourceASTokenEndpoint: in.ResourceASTokenEndpoint,
		IDJAG:                   idjag.AccessToken,
		Scopes:                  in.Scopes,
		HTTPClient:              in.HTTPClient,
	})
}

// ── helpers ──────────────────────────────────────────────────────────────────

func defaultClient(c *http.Client) *http.Client {
	if c != nil {
		return c
	}
	return &http.Client{Timeout: 10 * time.Second}
}

func parseOAuthError(status int, body []byte) error {
	var e OAuthError
	if err := json.Unmarshal(body, &e); err != nil || e.Code == "" {
		// Server returned non-JSON or non-OAuth shape. Wrap raw body.
		return fmt.Errorf("http %d: %s", status, strings.TrimSpace(string(body)))
	}
	e.HTTPStatus = status
	return &e
}
